src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs

```diff
@@ -93,27 +110,70 @@ private Block ProduceEmptyBlock(string payloadId, BlockHeader parentHeader, Payl
             return emptyBlock;
         }
 
-        private void ImproveBlock(string payloadId, BlockHeader parentHeader, PayloadAttributes payloadAttributes, Block emptyBlock)
+        private void ImproveBlock(string payloadId, BlockHeader parentHeader, PayloadAttributes payloadAttributes, Block currentBestBlock, DateTimeOffset startDateTime)
         {
-            if (_logger.IsTrace) _logger.Trace($"Start improving block from payload {payloadId} with parent {parentHeader}");
-            IBlockImprovementContext blockImprovementContext = _blockImprovementContextFactory.StartBlockImprovementContext(emptyBlock, parentHeader, payloadAttributes);
+            IBlockImprovementContext? oldContext = null;
+
+            _payloadStorage.AddOrUpdate(payloadId,
+                id => CreateBlockImprovementContext(id, parentHeader, payloadAttributes, currentBestBlock, startDateTime),
```

Comment:
`oldContext` is not set if the context is new. Is that intended? If so, why not call `Dispose` in line 128 so that you can remove a var?

Reply:
Right, I had to be focusing on other things.